### PR TITLE
parser: enable record types with optional fields

### PR DESCRIPTION
**Pull request by HoneyryderChuck**
this enables richer definitions of record types, which can only be defined nowadays with Hash[Symbol, untyped].

To handle #504

### CODE REVIEW

**Comment by HoneyryderChuck** [comment]
@soutaro this is a draft proposal, as in, the optional fields are parsed and stored, but not used yet. If you agree with this, I can proceed and use it in inferences (may need some help with that eventually as well).

**Comment by soutaro** [comment]
@HoneyryderChuck Thanks! Let's go with the syntax.

Having different hash tables for required and optional fields would result breaking the order of the records, for `Writer` or rewriting features. I'd like to add `#all_fields` attribute and let `#fields` and `#optional_fields` fields filtered from `#all_fields`.

```rbs
    class Record
      attr_reader fields: Hash[Symbol, [t, bool]]

      type loc = Location[bot, bot]

      def initialize: (fields: Hash[Symbol, t], location: loc?) -> void      # For compatibility with 3.4
                          | (all_fields: Hash[Symbol, [t, bool]], location: loc?) -> void

      include _TypeBase

      def fields: () -> Hash[Symbol, t]
      def optional_fields: () -> Hash[Symbol, t]?
    end
```

**Comment by HoneyryderChuck** [comment]
@soutaro can you check the changes? I believe this is what you were aiming for?

**Comment by soutaro** [comment]
Thanks @HoneyryderChuck! Looks great. Do you think it's ready to merge?

File: ext/rbs_extension/parser.c [hunk 1]
```diff
@@ -52,6 +52,11 @@ typedef struct {
   VALUE rest_keywords;
 } method_params;
 
+typedef struct {
```
Comment (soutaro) [inline ext/rbs_extension/parser.c:55]: Looks like the struct is no longer used.

File: sig/types.rbs [hunk 2]
```diff
@@ -301,16 +301,21 @@ module RBS
...
       def initialize: (fields: Hash[Symbol, t], location: loc?) -> void
+                    | (all_fields: Hash[Symbol, [t, bool]], location: loc?) -> void
 
       include _TypeBase
 
       attr_reader location: loc?
 
+      def fields: () -> Hash[Symbol, t]
+
+      def optional_fields: () -> Hash[Symbol, t]?
```
Comment (soutaro) [inline sig/types.rbs:317]: Ah, I made a mistake to put `?` at the end of the method type. It should return an empty hash.

```suggestion
      def optional_fields: () -> Hash[Symbol, t]
```

Comment (HoneyryderChuck) [reply sig/types.rbs:317]: if you look at the implementation, nothing is returned if all_fields is the same as fields. I think that's fine, i.e. we can avoid creating that extra hash when not necessary. Or do you see any use for it?

File: ext/rbs_extension/parser.c (hunk 1 above)
Comment (HoneyryderChuck) [reply ext/rbs_extension/parser.c]: thx, removed.

**Comment by HoneyryderChuck** [comment]
@soutaro anything on the above?

**Review by soutaro — APPROVED** [review]
Thanks! 👍

**Pull request merged**